docs/man/man3/MPI_Group.3
.\" DO NOT MODIFY THIS FILE! it was generated by rd2
.TH mpi_group.rd 1 "July 2001"
.SH MPI::Group
.PP
.SH Description
.PP
A wrapper class around MPI groups 
.SH Methods
.PP
.SH Constructors
.PP
.SH Informational Methods
.PP

.TP
.fi
.B
MPI::Group#size()
Returns the size of the group.

C equivalent: \&\fBMPI_Group_size()\fP


.TP
.fi
.B
MPI::Group#rank()
Returns the rank of the current process within the communicator.

C equivalent: \&\fBMPI_Comm_rank()\fP


.TP
.fi
.B
MPI::Group#translate_ranks(ary, grp)
Translates the group rankings listed in ary for the receiver group
into rankings in grp.  The new rankings are returned in an array, 
ordered respectively to ary.  If any of the rankings in ary 
do not correspond to a ranking in grp, the corresponding entry in
the return array is \&\fBMPI::UNDEFINED\fP.

C equivalent: \&\fBMPI_Group_translate_ranks()\fP


.TP
.fi
.B
MPI::Group#==(grp)
Returns true if the receiver and grp are precisely equal.

C equivalent: \&\fBMPI_Group_compare()\fP


.TP
.fi
.B
MPI::Group#similar?(grp)
Returns true if the receiver and grp are precisely similar.

C equivalent: \&\fBMPI_Group_compare()\fP
.SH Constructors
.PP

.TP
.fi
.B
MPI::Group#|(grp)
Returns the union of the receiver group and grp.

C equivalent: \&\fBMPI_Group_union()\fP


.TP
.fi
.B
MPI::Group#union(grp)
An alias for MPI::Group#|()

C equivalent: \&\fBMPI_Group_union()\fP


.TP
.fi
.B
MPI::Group#\-(grp)
Returns the difference of the receiver group and grp.

C equivalent: \&\fBMPI_Group_difference()\fP


.TP
.fi
.B
MPI::Group#difference(grp)
An alias for MPI::Group#\-()

C equivalent: \&\fBMPI_Group_difference()\fP


.TP
.fi
.B
MPI::Group#&(grp)
Returns the intersection of the receiver group and grp.

C equivalent: \&\fBMPI_Group_intersection()\fP


.TP
.fi
.B
MPI::Group#intersection(grp)
An alias for MPI::Group#&()

C equivalent: \&\fBMPI_Group_intersection()\fP


.TP
.fi
.B
MPI::Group#incl(ary)
Returns a new group with only those ranks listed in ary

C equivalent: \&\fBMPI_Group_incl()\fP


.TP
.fi
.B
MPI::Group#range_incl(ary)
Returns a new group with only those ranks in an array of ranges, 
ary.  ary is an array in which each element is an array of 
size three, describing a range.  The first element of one of these size 3
arrays is the start of the range, the second element is the end of the
range, and the third element is the stride taken between the range 
endpoints.  The stride may be negative, but not zero.  The ranks must be
distinct.

C equivalent: \&\fBMPI_Group_range_incl()\fP


.TP
.fi
.B
MPI::Group#excl(ary)
Returns a new group with only those ranks not listed in ary

C equivalent: \&\fBMPI_Group_excl()\fP


.TP
.fi
.B
MPI::Group#range_excl(ary)
Returns a new group with only those ranks not in an array of ranges, 
ary.  ary is an array in which each element is an array of 
size three, describing a range.  The first element of one of these size 3
arrays is the start of the range, the second element is the end of the
range, and the third element is the stride taken between the range 
endpoints.  The stride may be negative, but not zero.  The ranks must be
distinct.

C equivalent: \&\fBMPI_Group_range_excl()\fP



ChangeLog
Thu Jan 01 19:31:22 EST 2009
Aaron Bedra

Full merge of sourceforge codebase with other code-bases found on the internet.

Added some project maintainer notices to the code to point to me as the 
current maintainer.  

Mon Jun 30 14:06:27 CEST 2003
Rudi Cilibrasi

Fixed bug where mpiruby didn't search the Ruby library path for extension
modules.

Fixed bug with mpi_ruby crashing when programs finish.

Fixed bug where multiple send from slaves to a single master receiving
from anywhere could intermingle data length and data parts.

2001-07-27  Emil Ong  <onge@mentat.mcs.anl.gov>

	* examples/irecv.rb: Removed sleep from irecv example

	* configure: Removed configure

	* examples/Makefile.am: Added op example to Makefile

	* examples/redsubmit.rb, examples/op.rb, examples/redhalt.rb, examples/red.rb:
	New examples:
	User-defined operations
	Ruby Execution Daemon (red).

	* src/ops.rb, src/main.c, src/mpi.c, src/mpi_comm.c, src/mpi_group.c, src/mpi_keyval.c, src/mpi_op.c, src/mpi_op_fns.c, src/mpi_request.c:
	Fixed all of the rb_str_new2()'s that were causing marshalling problems.
	Fixed defines of singleton methods.
	Fixed dims_create()
	Now works with MPICH because of atexit(MPI_Finalize)
	Set the MPI error handler
	Fixed operators in MPI::Group (+ -> | and ^ -> &)

	* examples/Makefile.am: New examples.

	* docs/rd/mpi_group.rd, docs/rd/mpi_keyval.rd, docs/rd/mpi_op.rd, docs/rd/mpi_ruby.rd, docs/rd/mpi_comm.rd, docs/man/man3/Makefile.am, docs/man/man3/mpi_comm.3, docs/man/man3/MPI_Ruby.3, docs/man/man3/MPI_Status.3, docs/man/man3/MPI_Exception.3, docs/man/man3/MPI_Group.3, docs/man/man3/MPI_Keyval.3, docs/man/man3/MPI_Op.3, docs/man/man3/MPI_Request.3, docs/man/man3/MPI_Comm.3:
	Doc updates to reflect fixes to singleton methods in several classes.
	Typos fixed.

	* docs/man/man1/mpi_ruby.1, docs/man/man1/Makefile.am:
	Short doc on the interpreter itself (how to run)

	* docs/man/Makefile.am: Added man1

	* docs/html/index.html, docs/html/mpi_comm.html, docs/html/mpi_group.html, docs/html/mpi_keyval.html, docs/html/mpi_op.html, docs/html/mpi_ruby.html:
	Doc updates to reflect fixes to singleton methods in several classes.
	Typos fixed.

	* configure, configure.in: Bumped to 0.3
	Added man1/Makefile to output

2001-07-26  Emil Ong  <onge@mentat.mcs.anl.gov>

	* docs/rd/mpi_comm.rd, docs/rd/Makefile.am, docs/rd/mpi.rd:
	Added and fixed documentation
	Updated distribution system appropriately

	* docs/html/Makefile.am:
	Added the new documentation to the distribution system

	* docs/html/index.html: An index file for the html documentation

	* docs/html/mpi_comm.html, docs/html/mpi.html:
	Documentation fixes and updates

2001-07-25  Emil Ong  <onge@mentat.mcs.anl.gov>

	* ChangeLog, Makefile.in: Added a changelog

	* src/main.c, src/Makefile.in, src/mpi.c, src/mpi_comm.c, src/Makefile.am:
	Reverted to pointers for MPI_COMM_WORLD and MPI_COMM_SELF to avoid memory
	  freeing errors.
	Redeclared MPI module methods and module methods (duh!)
	Don't try to free WORLD and SELF
	Fixed issue with rb_str_new2() in Bcast

	* examples/basic.rb, examples/Makefile.am, examples/rpi.rb:
	new and fixed examples
	Added examples to the build path

	* docs/rd/mpi_keyval.rd, docs/rd/mpi_op.rd, docs/rd/mpi_request.rd, docs/rd/mpi_status.rd, docs/rd/Makefile.am, docs/rd/mpi_comm.rd, docs/rd/mpi_exception.rd, docs/rd/mpi_group.rd:
	Doc additions
	Documentation should be mostly done now

	* docs/man/man3/Makefile.am, docs/man/Makefile.am, docs/html/mpi_request.html, docs/html/mpi_status.html, docs/html/mpi_comm.html, docs/html/mpi_exception.html, docs/html/mpi_group.html, docs/html/mpi_keyval.html, docs/html/mpi_op.html, docs/html/Makefile.am:
	Doc additions

	* docs/Makefile.am: Doc addition

	* configure, configure.in, INSTALL, Makefile.am, Makefile.in, README, TODO, COPYRIGHT:
	Added docs and examples to make path
	Included COPYRIGHT
	Other various build fixes and configure cleanups

2001-07-20  Emil Ong  <onge@mentat.mcs.anl.gov>

	* src/mpi_comm.c, src/mpi_request.c, src/main.c, src/mpi.c:
	Fixed some errors with the constants
	Now assuming that MPI_Comm's are valid lvalues and rvalues

	* config.status: This never should have gone into CVS

	* docs/html/mpi_comm.html, docs/man/man3/mpi_comm.3, docs/MPI, docs/rd/mpi_comm.rd, docs/rd/mpi_request.rd, examples/allreduce.rb, examples/reduce.rb:
	Import of MPI_Ruby


	* docs/html/mpi_comm.html, docs/man/man3/mpi_comm.3, docs/MPI, docs/rd/mpi_comm.rd, docs/rd/mpi_request.rd, examples/allreduce.rb, examples/reduce.rb:
	New file.

	* examples/allgather.rb, examples/basic.rb, examples/bcast.rb, examples/gather.rb, examples/hello.rb, examples/irecv.rb, src/Makefile.am, src/Makefile.in, src/mpi_exception.c, src/mpi_exception.h, src/mpi_group.h, src/mpi_keyval.c, src/mpi_keyval.h, src/mpi_op.c, src/mpi_op.h, src/ops.rb:
	Import of MPI_Ruby


	* examples/allgather.rb, examples/basic.rb, examples/bcast.rb, examples/gather.rb, examples/hello.rb, examples/irecv.rb, src/Makefile.am, src/Makefile.in, src/mpi_exception.c, src/mpi_exception.h, src/mpi_group.h, src/mpi_keyval.c, src/mpi_keyval.h, src/mpi_op.c, src/mpi_op.h, src/ops.rb:
	New file.

	* config.guess, config.status, config.sub, configure, depcomp, mkinstalldirs, src/global.h, src/main.c, src/mpi.c, src/mpi_comm.c, src/mpi_comm.h, src/mpi_group.c, src/mpi_op_fns.c, src/mpi_request.c, src/mpi_request.h, src/mpi_status.c, src/mpi_status.h, stamp-h.in:
	Import of MPI_Ruby


	* config.guess, config.status, config.sub, configure, depcomp, mkinstalldirs, src/global.h, src/main.c, src/mpi.c, src/mpi_comm.c, src/mpi_comm.h, src/mpi_group.c, src/mpi_op_fns.c, src/mpi_request.c, src/mpi_request.h, src/mpi_status.c, src/mpi_status.h, stamp-h.in:
	New file.

	* aclocal.m4, config.h.in, configure.in, install-sh, Makefile.am, Makefile.in, missing, TODO:
	Import of MPI_Ruby


	* aclocal.m4, config.h.in, configure.in, install-sh, Makefile.am, Makefile.in, missing, TODO:
	New file.



src/mpi_group.c
#include "mpi.h"
#include "ruby.h"
#include "global.h"

VALUE cGroup;

static void group_free(MPI_Group *grp)
{
    int rv;

    rv = MPI_Group_free(grp);
    mpi_exception(rv);
    free(grp);
}

VALUE group_new(MPI_Group *grp)
{
    VALUE tdata;

    tdata = Data_Wrap_Struct(cGroup, NULL, group_free, grp);
    rb_obj_call_init(tdata, 0, NULL);

    return tdata;
}

MPI_Group *group_get_mpi_group(VALUE rgrp)
{
    MPI_Group *grp;

    Data_Get_Struct(rgrp, MPI_Group, grp);

    return grp;
}

static VALUE group_size(VALUE self)
{
    int rv, size;
    MPI_Group *grp;

    Data_Get_Struct(self, MPI_Group, grp);

    rv = MPI_Group_size(*grp, &size);
    mpi_exception(rv);

    return rb_fix_new(size);
}

static VALUE group_rank(VALUE self)
{
    int rv, rank;
    MPI_Group *grp;

    Data_Get_Struct(self, MPI_Group, grp);

    rv = MPI_Group_rank(*grp, &rank);
    mpi_exception(rv);

    if (rank == MPI_UNDEFINED)
        return UNDEFINED;

    return rb_fix_new(rank);
}

static VALUE group_translate_ranks(VALUE self, VALUE ary, VALUE rgrp2)
{
    VALUE outary;
    MPI_Group *grp1, *grp2;
    int rv, i, len, *ranks1, *ranks2;

    Data_Get_Struct(self, MPI_Group, grp1);
    Data_Get_Struct(grp2, MPI_Group, grp2);

    len = RARRAY(ary)->len;
    ranks1 = ALLOCA_N(int, len);
    ranks2 = ALLOCA_N(int, len);

    for (i = 0; i < len; i++)
        ranks1[i] = FIX2INT(rb_ary_entry(ary, i));

    rv = MPI_Group_translate_ranks(*grp1, len, ranks1, *grp2, ranks2);
    mpi_exception(rv);

    outary = rb_ary_new2(len);
    for (i = 0; i < len; i++) {
        if (ranks2[i] == MPI_UNDEFINED)
            rb_ary_store(outary, i, UNDEFINED);
        else
            rb_ary_store(outary, i, rb_fix_new(ranks2[i]));
    }

    return outary;
}

static VALUE group_equal(VALUE self, VALUE rgrp2)
{
    int rv, flag;
    MPI_Group *grp1, *grp2;

    Data_Get_Struct(self, MPI_Group, grp1);
    Data_Get_Struct(grp2, MPI_Group, grp2);

    rv = MPI_Group_compare(*grp1, *grp2, &flag);
    mpi_exception(rv);

    return flag == MPI_IDENT ? Qtrue : Qfalse;
}

static VALUE group_similar_p(VALUE self, VALUE rgrp2)
{
    int rv, flag;
    MPI_Group *grp1, *grp2;

    Data_Get_Struct(self, MPI_Group, grp1);
    Data_Get_Struct(grp2, MPI_Group, grp2);

    rv = MPI_Group_compare(*grp1, *grp2, &flag);
    mpi_exception(rv);

    return flag == MPI_SIMILAR ? Qtrue : Qfalse;
}

static VALUE group_union(VALUE self, VALUE rgrp2)
{
    int rv;
    MPI_Group *grp1, *grp2, *newgrp;

    Data_Get_Struct(self, MPI_Group, grp1);
    Data_Get_Struct(grp2, MPI_Group, grp2);

    newgrp = ALLOC(MPI_Group);

    rv = MPI_Group_union(*grp1, *grp2, newgrp);
    mpi_exception(rv);

    return group_new(newgrp);
}

static VALUE group_intersection(VALUE self, VALUE rgrp2)
{
    int rv;
    MPI_Group *grp1, *grp2, *newgrp;

    Data_Get_Struct(self, MPI_Group, grp1);
    Data_Get_Struct(grp2, MPI_Group, grp2);

    newgrp = ALLOC(MPI_Group);

    rv = MPI_Group_intersection(*grp1, *grp2, newgrp);
    mpi_exception(rv);

    return group_new(newgrp);
}

static VALUE group_difference(VALUE self, VALUE rgrp2)
{
    int rv;
    MPI_Group *grp1, *grp2, *newgrp;

    Data_Get_Struct(self, MPI_Group, grp1);
    Data_Get_Struct(grp2, MPI_Group, grp2);

    newgrp = ALLOC(MPI_Group);

    rv = MPI_Group_difference(*grp1, *grp2, newgrp);
    mpi_exception(rv);

    return group_new(newgrp);
}

static VALUE group_incl(VALUE self, VALUE ary)
{
    int rv, i, len, *ranks;
    MPI_Group *grp, *newgrp;

    Data_Get_Struct(self, MPI_Group, grp);

    newgrp = ALLOC(MPI_Group);

    len = RARRAY(ary)->len;
    ranks = ALLOCA_N(int, len);

    for (i = 0; i < len; i++)
        ranks[i] = FIX2INT(rb_ary_entry(ary, i));

    rv = MPI_Group_incl(*grp, len, ranks, newgrp);
    mpi_exception(rv);

    return group_new(newgrp);
}

static VALUE group_excl(VALUE self, VALUE ary)
{
    int rv, i, len, *ranks;
    MPI_Group *grp, *newgrp;

    Data_Get_Struct(self, MPI_Group, grp);

    newgrp = ALLOC(MPI_Group);

    len = RARRAY(ary)->len;
    ranks = ALLOCA_N(int, len);

    for (i = 0; i < len; i++)
        ranks[i] = FIX2INT(rb_ary_entry(ary, i));

    rv = MPI_Group_excl(*grp, len, ranks, newgrp);
    mpi_exception(rv);

    return group_new(newgrp);
}

static VALUE group_range_incl(VALUE self, VALUE ary)
{
    int rv, i, len, **ranks;
    MPI_Group *grp, *newgrp;

    Data_Get_Struct(self, MPI_Group, grp);

    newgrp = ALLOC(MPI_Group);

    len = RARRAY(ary)->len;
    ranks = ALLOCA_N(int *, len);

    for (i = 0; i < len; i++) {
        int j;
        VALUE range;
        
        range = rb_ary_entry(ary, i);
        ranks[i] = ALLOCA_N(int, 3);
        
        for (j = 0; j < 3; j++)
            ranks[i][j] = FIX2INT(rb_ary_entry(range, j));
    }

    /* Thank you, cdecl */
    rv = MPI_Group_range_incl(*grp, len, (int (*)[3])ranks, newgrp);
    mpi_exception(rv);

    return group_new(newgrp);
}

static VALUE group_range_excl(VALUE self, VALUE ary)
{
    int rv, i, len, **ranks;
    MPI_Group *grp, *newgrp;

    Data_Get_Struct(self, MPI_Group, grp);

    newgrp = ALLOC(MPI_Group);

    len = RARRAY(ary)->len;
    ranks = ALLOCA_N(int *, len);

    for (i = 0; i < len; i++) {
        int j;
        VALUE range;
        
        range = rb_ary_entry(ary, i);
        ranks[i] = ALLOCA_N(int, 3);
        
        for (j = 0; j < 3; j++)
            ranks[i][j] = FIX2INT(rb_ary_entry(range, j));
    }

    /* Thank you, cdecl */
    rv = MPI_Group_range_excl(*grp, len, (int (*)[3])ranks, newgrp);
    mpi_exception(rv);

    return group_new(newgrp);
}

void Init_Group()
{
    cGroup = rb_define_class_under(mMPI, "Group", rb_cObject);

    rb_define_method(cGroup, "size", group_size, 0);
    rb_define_method(cGroup, "rank", group_rank, 0);

    rb_define_method(cGroup, "translate_ranks", group_translate_ranks, 2);

    rb_define_method(cGroup, "==", group_equal, 1);
    rb_define_method(cGroup, "similar?", group_similar_p, 1);

    rb_define_method(cGroup, "|", group_union, 1);
    rb_define_method(cGroup, "union", group_union, 1);
    rb_define_method(cGroup, "-", group_difference, 1);
    rb_define_method(cGroup, "difference", group_difference, 1);
    rb_define_method(cGroup, "&", group_intersection, 1);
    rb_define_method(cGroup, "intersection", group_intersection, 1);

    rb_define_method(cGroup, "incl", group_incl, 1);
    rb_define_method(cGroup, "range_incl", group_range_incl, 1);
    rb_define_method(cGroup, "excl", group_excl, 1);
    rb_define_method(cGroup, "range_excl", group_range_excl, 1);
}
